Serialize data docs and add data doc import

megcrenshaw opened the pull request:
This gets the data doc import ready by (a) adding the import, (b) writing the serialization, and then (c) testing locally both the import and the serialization. We'll need to work with the Curriculum Team before actually running the import on levelbuilder.

There are 103 data docs currently on Curriculum Builder. Once I ran the script locally, I saw 103 data docs created:

<img width="916" alt="Screen Shot 2022-09-22 at 1 08 22 PM" src="https://user-images.githubusercontent.com/9142121/191829363-3832a69c-47bb-463b-945d-22504a5c558f.png">

There were also 103 json files written:

<img width="1003" alt="Screen Shot 2022-09-22 at 12 58 08 PM" src="https://user-images.githubusercontent.com/9142121/191829189-10a4806d-9eac-4901-b072-1fd609567e3d.png">

Here's a sample of their content:

<img width="941" alt="Screen Shot 2022-09-22 at 12 59 04 PM" src="https://user-images.githubusercontent.com/9142121/191829319-ba32f270-d4f7-4ad5-8d15-9bd74f367074.png">



## Links

<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story

I added some unit tests, but I think the model tests can get stronger once the seeding logic is added. I didn't do that here.

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

megcrenshaw commented on bin/curriculum/import_data_docs.rb:
```diff
@@ -0,0 +1,139 @@
+    while data_doc_key_set.include?(unique_key)
+      unique_key = "#{key_base}-#{iteration}"
+      iteration += 1
+    end
```
This is catching for duplicate keys –– it's not necessary for data_docs (since all of them have the URL of `\data-library\[key]`, but I kept it in in case we want that logic later on

davidsbailey replied:
again I would suggest removing code that you don't need. you will know where to find it when you need it

megcrenshaw commented on dashboard/test/models/data_doc_test.rb:
```diff
@@ -13,4 +13,18 @@ class DataDocTest < ActiveSupport::TestCase
     end
   end
+
+  test "serialize returns an object with the correct fields" do
```
This is kinda a silly test ... I'm thinking I can replace this by using seeding once that logic is implemented

davidsbailey replied:
makes sense... running both serialize and seed is a nice way to test things

davidsbailey commented on dashboard/app/models/data_doc.rb:
```diff
@@ -19,4 +19,29 @@ class DataDoc < ApplicationRecord
   validate :validate_key_format
+
+  def file_path
+    Rails.root.join("config/data_docs/#{key.parameterize(preserve_case: false)}.json")
```
super tiny nit: `preserve_case: false` looks like the default, so no need to specify it

davidsbailey commented on bin/curriculum/import_data_docs.rb:
```diff
@@ -0,0 +1,139 @@
+    current_slug = data_doc_slug_stack.pop
+
+    # only import Data Docs
+    # next unless current_slug.start_with?('concepts/data-library')
```
from the code above, it does not look like you will ever want to enable this right? if so, better to delete it than to comment it out.

megcrenshaw replied:
Yikes! Good call cleaning this up –– just simplified things based on the data-library simplicity

davidsbailey reviewed (approved):
nice progress! lgtm after addressing PR comments.

You may find that you need to make updates to this script as you continue building on this feature. I would suggest doing a more thorough check a bit later, when you are almost ready to do the import, to make sure you've got everything exactly how you want it here.

megcrenshaw commented:
> nice progress! lgtm after addressing PR comments.
> 
> You may find that you need to make updates to this script as you continue building on this feature. I would suggest doing a more thorough check a bit later, when you are almost ready to do the import, to make sure you've got everything exactly how you want it here.

Good call. I'll probably inquire at a checkin about what things are most often "weird" or missed, but I also think having the page that shows all the docs first will be helpful for testing!
